### PR TITLE
Java: Improve virtual dispatch via better unification check and deduplicate code with parameterised module

### PR DESCRIPTION
**Pull request by aschackmull**
The first commit is a pure refactor that changes all four copies of the unification code to instead use a single parameterised module.
The second commit improves the unification algorithm to account for the upper bounds of type variables and wildcards - until now the code assumed that all type variables and wildcards were merely `extends Object` and thus unifiable with any `RefType`.
The third commit also accounts for wildcards with lower bounds.

### CODE REVIEW

**Comment by aschackmull** [comment]
cc @ginsbach to check parameterised module usage.

**Comment by ginsbach** [comment]
Looks good to me from a parameterised modules perspective.

**Comment by aschackmull** [comment]
Finally managed to get a decent-looking dca run.

File: java/ql/lib/semmle/code/java/dispatch/Unification.qll [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        not other.getASourceSupertype*() = upperbound
+      )
+      or
+      exists(RefType upperbound1, RefType upperbound2 |
+        t1.(BoundedType).getAnUltimateUpperBoundType() = upperbound1 and
+        t2.(BoundedType).getAnUltimateUpperBoundType() = upperbound2 and
+        notHaveIntersection(upperbound1, upperbound2)
+      )
+      or
+      exists(RefType lowerbound, RefType upperbound |
```
Comment (smowton) [inline java/ql/lib/semmle/code/java/dispatch/Unification.qll:93]:
```suggestion
      // Can't unify `? super X` against `Y` or `? extends Y` where `Y` isn't an ancestor of `X`:
      exists(RefType lowerbound, RefType upperbound |
```

File: java/ql/lib/semmle/code/java/dispatch/Unification.qll [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        not lowerbound instanceof BoundedType
+        or
+        t2.(Wildcard).getLowerBoundType().(RefType).getSourceDeclaration() = lowerbound and
+        getUpperBound(t1).getSourceDeclaration() = upperbound and
+        not lowerbound instanceof BoundedType
+      |
+        not lowerbound.getASourceSupertype*() = upperbound
+      )
+      or
+      exists(BoundedType lowerbound, RefType upperbound |
```
Comment (smowton) [inline java/ql/lib/semmle/code/java/dispatch/Unification.qll:105]:
```suggestion
      // Can't unify `? super T`, where `T` is a type variable `T extends S`, with a type that doesn't intersect with `S`:
      exists(BoundedType lowerbound, RefType upperbound |
```

File: java/ql/lib/semmle/code/java/dispatch/Unification.qll [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      exists(RefType lowerbound, RefType upperbound |
+        t1.(Wildcard).getLowerBoundType().(RefType).getSourceDeclaration() = lowerbound and
+        getUpperBound(t2).getSourceDeclaration() = upperbound and
+        not lowerbound instanceof BoundedType
+        or
+        t2.(Wildcard).getLowerBoundType().(RefType).getSourceDeclaration() = lowerbound and
+        getUpperBound(t1).getSourceDeclaration() = upperbound and
+        not lowerbound instanceof BoundedType
+      |
+        not lowerbound.getASourceSupertype*() = upperbound
```
Comment (smowton) [inline java/ql/lib/semmle/code/java/dispatch/Unification.qll:102]:
```suggestion
        or
        t2.(Wildcard).getLowerBoundType().(RefType).getSourceDeclaration() = lowerbound and
        getUpperBound(t1).getSourceDeclaration() = upperbound and
      |
        not lowerbound instanceof BoundedType
        not lowerbound.getASourceSupertype*() = upperbound
```

**Review by smowton — APPROVED** [review]
Some optional comments and factoring a duplicated line, otherwise LGTM

**Comment by aschackmull** [comment]
> Some optional comments and factoring a duplicated line

Fixed.

File: java/ql/lib/semmle/code/java/dispatch/Unification.qll [hunk 4]
```diff
@@ -0,0 +1,136 @@
+/**
+ * Provides a module to check whether two `ParameterizedType`s are unifiable.
```
Comment (hvitved) [inline java/ql/lib/semmle/code/java/dispatch/Unification.qll:2]: I would probably make this an internal module, by moving inside of an `internal` folder.

**Review by hvitved — APPROVED** [review]

**Pull request merged**